sdk/src/main/kotlin/io/provenance/scope/sdk/Client.kt

```diff
@@ -110,19 +136,41 @@ class Client(val inner: SharedClient, val affiliate: Affiliate) {
             .also { it.client = this } // TODO remove when class is moved over
             .setContractSpec(contractSpec)
             .setProvenanceReference(contractRef)
-            .addParticipant(affiliate.partyType, affiliate.encryptionKeyRef.publicKey.toPublicKeyProtoOS())
+            .addParticipant(affiliate.partyType, affiliate.encryptionKeyRef.publicKey.toPublicKeyProto())
     }
 
     fun execute(session: Session, affiliateSharePublicKeys: Collection<PublicKey> = listOf()): ExecutionResult {
         val input = session.packageContract()
         val result = inner.contractEngine.handle(affiliate.encryptionKeyRef, affiliate.signingKeyRef, input, session.scope, affiliateSharePublicKeys)
 
-        return when (result.isSigned(session.scope, inner.config.mainNet)) {
+        return when (result.isSigned(inner.config.mainNet)) {
             true -> SignedResult(session, result, inner.config.mainNet) // todo: better way to get the scope/session, we will always need some minimal info for creating a new scope if not existant
-            false -> throw NotImplementedError("Multi-party contract support not yet implemented")
+            false -> FragmentResult(input, result) // todo: do we need both input and result?
         }
     }
 
+    fun registerMailHandler(handler: MailHandlerFn) {
```

Comment:
This looks optimal for kotlin, but for Java it might be awkward to call this function with a lambda. Do you think it makes sense to add a second function that does the same thing and takes in a class with a single function so it's like the java callback style?

Reply:
Also, what if this function took in an executor to use (thread pool). That way we would simplify the logic in the companion object, also it would give the caller ownership of the executor so they could shut it down when they wanted?

Reply:
I think both of those make sense, I might try using the Java SAM Function1 interface and see if that works to accept a regular Kotlin lambda as it is supposed to now according to this thread https://discuss.kotlinlang.org/t/under-what-conditions-can-i-call-a-sam-accepting-method-using-a-trailing-lambda/15238 . As for the thread pool, I think I could just require them to pass in a scheduled executor and use that... maybe I should just return the ScheduledFuture as well so they can manage when that particular listener is un-registered as well?

Reply:
if we were really nice we could provide a default that would be handled like it currently is...

Reply:
@scirner22 I made these changes, let me know what you think. This allows us to pass either a lambda like before, or some object implementing Function<MailboxEvent, Boolean>

Reply:
👍 we can roll with it and see what it looks like in Kotlin and Java examples.